src/plugins/part.c

```diff
@@ -1317,107 +1254,354 @@ static PedPartition* add_part_to_disk (PedDevice *dev, PedDisk *disk, BDPartType
+           current maxal number of partitions suppported by the label
+           is the next (logical) partition number */
+        partno = fdisk_get_npartitions (cxt);
+
+    } else {
+        status = fdisk_partition_next_partno (npa, cxt, &partno);
+        if (status != 0) {
+            g_set_error (error, BD_PART_ERROR, BD_PART_ERROR_FAIL,
+                         "Failed to get new partition number");
+            fdisk_unref_partition (npa);
+            close_context (cxt);
             bd_utils_report_finished (progress_id, (*error)->message);
             return NULL;
         }
+    }
 
-        if (ped_part->type == PED_PARTITION_EXTENDED) {
-            /* can at minimal start where the first logical partition can start - the start of the extended partition + 1 MiB aligned up */
-            if (start < ((ped_part->geom.start * dev->sector_size) + 1 MiB + dev->sector_size - 1))
-                start = (ped_part->geom.start * dev->sector_size) + 1 MiB + dev->sector_size - 1;
-        } else {
-            /* can at minimal start where the next logical partition can start - the end of the previous partition + 1 MiB aligned up */
-            if (start < ((ped_part->geom.end * dev->sector_size) + 1 MiB + dev->sector_size - 1))
-                start = (ped_part->geom.end * dev->sector_size) + 1 MiB + dev->sector_size - 1;
-        }
+    status = fdisk_partition_set_partno (npa, partno);
+    if (status != 0) {
+        g_set_error (error, BD_PART_ERROR, BD_PART_ERROR_FAIL,
+                     "Failed to set new partition number");
+        fdisk_unref_partition (npa);
+        close_context (cxt);
+        bd_utils_report_finished (progress_id, (*error)->message);
+        return NULL;
     }
 
-    ped_part = add_part_to_disk (dev, ped_disk, type, start, size, align, error);
-    if (!ped_part) {
-        /* error is already populated */
-        ped_disk_destroy (ped_disk);
-        ped_device_destroy (dev);
+    status = fdisk_add_partition (cxt, npa, NULL);
+    if (status != 0) {
+        g_set_error (error, BD_PART_ERROR, BD_PART_ERROR_FAIL,
+                     "Failed to add new partition to the table: %s", strerror_l (-status, c_locale));
+        fdisk_unref_partition (npa);
+        close_context (cxt);
         bd_utils_report_finished (progress_id, (*error)->message);
         return NULL;
     }
 
-    succ = disk_commit (ped_disk, disk, error);
-    if (succ)
-        ret = get_part_spec (dev, ped_disk, ped_part, error);
+    if (!write_label (cxt, disk, error)) {
+        bd_utils_report_finished (progress_id, (*error)->message);
+        fdisk_unref_partition (npa);
+        close_context (cxt);
+        return NULL;
+    }
 
-    /* the partition gets destroyed together with the disk*/
-    ped_disk_destroy (ped_disk);
-    ped_device_destroy (dev);
+    /* the in-memory model of the new partition is not updated, we need to
+       read the spec manually */
+    if (fdisk_partition_has_partno (npa)) {
```

Comment:
So what is reading the spec here?

Reply:
I've moved the comment to the line where we get the spec.